### PR TITLE
Updates for 0.4.0

### PR DESCRIPTION
**Pull request by mmellado**
- Fixed state reassignment for state variables that are dependant on props
- Added support for additional valid props to form field components
- Added `type="button"` to the toggle password button in the TextField component

### CODE REVIEW

File: TextField/TextField.js [hunk 1]
```diff
@@ -149,21 +166,32 @@ export default class TextField extends Component {
...
-      if (maxContentLength && this.state.textFieldValue.length > maxContentLength) {
-        this.setState({
-          isMaxLengthExceeded: true,
-        }, () => {
-          if (this.props.onChange) {
-            this.props.onChange(e, new Error(this.maxLengthError));
-          }
-        });
-      } else {
+      this.attemptOnChangeCallback({
```
Comment (elvisvoer) [inline TextField/TextField.js]: I cannot leave a comment in line 164 but we should remove `e.persist()` as it is not needed anymore

File: TextArea/TextArea.js [hunk 2]
```diff
@@ -101,36 +102,85 @@ export default class TextArea extends Component {
...
-        this.setState({
-          isMaxLengthExceeded: true,
-        }, () => {
-          if (this.props.onChange) {
-            this.props.onChange(e, new Error(this.maxLengthError));
-          }
-        });
-      } else {
+      this.attemptOnChangeCallback({
+        value: this.state.textAreaValue,
```
Comment (elvisvoer) [inline TextArea/TextArea.js]: if we don't return an object in the form `{target: {value: myvalue}}` we break the onChange callback API so we should reflect this in the changelog

**Comment by alexjeffburke** [comment]
I've pondered this a little - to make this more of a drop-in for what we have now (since this is also a bugfix release that we could benefit from now), how about we temporarily duplicate the value sent back as `{ target: { value: <theValue> } }`. Since executing onChange is unified in one place we can put a big comment above it and remove it in 0.5.0. Just a thought.

**Comment by mmellado** [comment]
I agree on the e.target.value, will update to go inline with this structure. I don't think that it needs to be removed from 0.5.0, the question now is if we should standardize this for all inputs. I'm still concerned that things like the keystroke etc are now lost when the user types on an input field. How about we go back to an event and send back an object that respects the event structure for the onPropChange callbacks??

**Comment by mmellado** [comment]
After some discussion, we will not trigger the onChange event at componentWillReceiveProps, as this is not a native onChange event. If this becomes a need in the future, we will explore the best way of standardizing this behavior

File: TextArea/TextArea.js [hunk 3]
```diff
@@ -101,36 +102,79 @@ export default class TextArea extends Component {
...
+    }
+  }
+
+  /**
    * Takes care of updating the value of the input field when the user types in it
    *
    * @param {Event} [e] - The event triggered when the user types
    */
   handleInputValue = e => {
-    e.persist();
```
Comment (elvisvoer) [inline TextArea/TextArea.js]: We need to keep `e.persist()` here

**Pull request merged**